### PR TITLE
etcdserver: Fix migrations

### PR DESCRIPTION
**Pull request by barakmich**
Actually having a hook for migrations is useful. We have one, let's use it.

### CODE REVIEW

File: etcdserver/server.go [hunk 1]
```diff
@@ -146,16 +146,20 @@ func NewServer(cfg *ServerConfig) (*EtcdServer, error) {
 	var n raft.Node
 	var s *raft.MemoryStorage
 	var id types.ID
-	walVersion, err := wal.DetectVersion(cfg.DataDir)
+	var ss *snap.Snapshotter
+	baseDataDir := cfg.DataDir
+	// Data dir lives under member now.
+	cfg.DataDir = path.Join(cfg.DataDir, "member")
```
Comment (yichengq) [inline etcdserver/server.go]: I don't like this way.
We have function cfg.SnapDir(), which relies on the value of cfg.DataDir. And it cannot behave correctly until you set the value here. It is uncontrollable.

**Comment by xiang90** [comment]
@barakmich Can you please change the short commit message to follow the format `pkg: message`?

File: etcdserver/server.go (hunk 1 above)
Comment (xiang90) [inline etcdserver/server.go]: The root problem: What should be the data-dir in etcdserver.Config. 

Should it be the base dir (without member)? Or should it be the member dir? 
I am OK with both, but prefer member dir.

In my opinion whatever we choose, we should make it consistent throughout the code inside etcdserver pkg. In another word, cfg.DataDir should be readonly.

**Comment by barakmich** [comment]
Squashed, and if you're okay with both (I say it should be the base dir) then it's ready to go.

**Comment by xiang90** [comment]
LGTM if it works.

**Comment by yichengq** [comment]
lgtm

**Comment by barakmich** [comment]
Cool. And yeah, I triple checked upgrading data dirs simply.

**Pull request merged**